sdlib/d/gc/threadpool.d

```diff
@@ -0,0 +1,273 @@
+		detachSelf();
+		auto pool = cast(shared(ThreadPool*)) ctx;
+		pool.executeWork(false);
+		return null;
+	}
+
+public:
+
+	/**
+	 * Start threads using the given thread buffer. All threads will wait
+	 * for work dispatched to them using the `dispatch` function. Optimal
+	 * thread buffer size should be number of cores, or number of cores - 1
+	 * if you intend to use the calling thread for work.
+	 *
+	 * Note that these threads will be marked as detached, so they will not
+	 * be scanned if a GC cycle is run.
+	 */
+	void startThreads(pthread_t[] threadBuffer) shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(ThreadPool*) &this).startThreadsImpl(threadBuffer);
+	}
+
+	/**
+	 * Dispatch work to the threads. `count` instances of the `work`
+	 * delegate are run, with each call to the delegate receiving an index
+	 * from `0` to `count - 1`. If `count` is greater than the number of
+	 * threads in the pool, threads will re-call the delegate, until
+	 * `count` calls to the delegate are made. No ordering of the index is
+	 * guaranteed, but it is guaranteed that only one instance of an index
+	 * will be used.
+	 */
+	void dispatch(void delegate(uint) work, uint count) shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(ThreadPool*) &this).dispatchImpl(work, count);
+	}
+
+	/**
+	 * Wait for all work to be complete. `useThisThread` set to true will
+	 * use the calling thread to also do work if needed.
+	 */
+	void waitForIdle(bool useThisThread = true) shared {
+		if (useThisThread) {
+			executeWork(true);
+		}
+
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		auto tp = cast(ThreadPool*) &this;
+		mutex.waitFor(tp.allThreadsIdle);
+	}
+
+	/**
+	 * Stop all threads, and join them. If you need to start the thread
+	 * pool up again, you can call `startThreads` again.
+	 *
+	 * Threads are not stopped until all scheduled work is completed.
+	 *
+	 * `useThisThread` = true will use the calling thread to do some work
+	 * if available.
+	 */
+	void joinAll(bool useThisThread = true) shared {
+		if (useThisThread) {
+			executeWork(true);
+		}
+
+		{
+			mutex.lock();
+			scope(exit) mutex.unlock();
+
+			(cast(ThreadPool*) &this).stopAllThreads();
+		}
+
+		/**
+		  * This is called without the lock held, and not in a shared
+		  * method, because we are exiting all threads, and we don't
+		  * want to impede the message of the exitFlag.
+		  */
+		(cast(ThreadPool*) &this).joinAllThreadsImpl();
+	}
+
+private:
+	void startThreadsImpl(pthread_t[] threadBuffer) {
+		assert(threads.length == 0);
+
+		threads = threadBuffer;
+		exitFlag = false;
+		activeThreads = cast(uint) threadBuffer.length;
+		foreach (ref tid; threads) {
+			pthread_create(&tid, null, threadEntry, cast(void*) &this);
+		}
+	}
+
+	void dispatchImpl(void delegate(uint) work, uint count) {
+		assert(mutex.isHeld(), "Mutex not held!");
+		auto sharedThis = cast(shared(ThreadPool)*) &this;
+		sharedThis.mutex.waitFor(noMoreWork);
```

Comment:
This was puzzling for me. `Mutex.waitFor` requires a shared mutex, yet the mutex must be locked to call it. Typically, the mutex is unshared in a non-shared method, so this was awkward. What is the reason for `waitFor` being shared-only?

You see an interesting thing here as well: https://github.com/snazzy-d/sdc/blob/f961950b1dce0026b11ecdc7ac0ebb386486572d/sdlib/d/gc/scanner.d#L173

There, you are in a shared method, but you need to call an unshared method because the thing is locked at the time.

I suppose I could mark the mutex itself shared, so it's always compatible?